streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java

```diff
@@ -723,15 +735,61 @@ public GroupAssignment assign(final Cluster metadata, final GroupSubscription gr
         return taskIdsForConsumerAssignment;
     }
 
-    private void upgradeSubscriptionVersionIfNeeded(final int leaderSupportedVersion) {
-        if (leaderSupportedVersion > usedSubscriptionMetadataVersion) {
-            log.info("Sent a version {} subscription and group leader's latest supported version is {}. " +
-                         "Upgrading subscription metadata version to {} for next rebalance.",
+    private void validateMetadataVersions(final int receivedAssignmentMetadataVersion,
+                                          final int latestCommonlySupportedVersion) {
+
+        if (receivedAssignmentMetadataVersion > usedSubscriptionMetadataVersion) {
+            log.error("Leader sent back an assignment with version {} which was greater than our used version {}",
+                receivedAssignmentMetadataVersion, usedSubscriptionMetadataVersion);
+            throw new IllegalStateException(
+                "Sent a version " + usedSubscriptionMetadataVersion
+                    + " subscription but got an assignment with higher version "
+                    + receivedAssignmentMetadataVersion + "."
+            );
+        }
+
+        if (latestCommonlySupportedVersion > LATEST_SUPPORTED_VERSION) {
+            log.error("Leader sent back assignment with commonly supported version {} that is greater than our"
+                + "actual latest supported version {}", latestCommonlySupportedVersion, LATEST_SUPPORTED_VERSION);
+            throw new IllegalStateException("Can't upgrade to metadata version greater than we support");
+        }
+    }
+
+    // Returns true if subscription version was changed, indicating version probing and need to rebalance again
+    private boolean maybeUpdateSubscriptionVersion(final int receivedAssignmentMetadataVersion,
+                                                   final int latestCommonlySupportedVersion) {
+
+        if (receivedAssignmentMetadataVersion >= EARLIEST_PROBEABLE_VERSION) {
```

Comment:
If receivedAssignmentMetadataVersion < EARLIEST_PROBEABLE_VERSION, it is also a pretty bad situation since it means that the leader is too old to even be covered with version probing --- in fact, if it is that old, upon getting the new version it would throw exception and die immediately, so we'd not even see the assignment back at all --- so we'd better also log it as an ERROR to complete the logic.

Reply:
True, we should actually check for this (in `validateMetadataVersions`) and throw an exception -- it basically means someone tried to do a rolling bounce from < 2.0 without setting the `UPGRADE_FROM` config

Reply:
SG.

Reply:
Actually, this would not be an error but possibly just an upgrade from a very old version -- see https://github.com/apache/kafka/pull/7436